### DISCUSSION TITLE
[Question][Hw6] About IOOracle API

### DISCUSSION THREAD

**Discussion by kohs100** [prosyslab-classroom/cs424-program-reasoning#303] [Q&A]
It seems that since this assignment utilizes functional constraints, IOOracle API won't work like past assignments.
Is there other way to evaluate constraints with specific input memory without using Spec.verify (=not using Z3)?

**Comment by sujin0529** [comment]
Hi,

If "specific input memory" means "input and output example", then there isn't a way to evaluate constraints without using Z3.

Thanks.

**Comment by kohs100** [comment]
> Hi,
> 
> If "specific input memory" means "input and output example", then there isn't a way to evaluate constraints without using Z3.
> 
> Thanks.

Before the verify process, I want to prevent creating duplicated counterexamples by pre-evaluating newly synthesized terminals by using the counterexamples given in the previous verification. If there is no way to evaluate constraints without using Z3, then is this optimization impossible(or, unnecessary)?

**Comment by kohs100** [comment]
I found that I can implement such optimization with spec.constraints list. Thanks!
